### ISSUE TITLE
Docs on sharing custom hooks with React

### ISSUE THREAD

**Issue by justinfagnani** [matthewp/haunted#130]
I've had a few people ask about hooks for LitElement because they wanted to share custom hooks with their React codebase. #87 is a great start, but considering that most custom hooks import the built-in hooks directly from a React module, the path to code-sharing still isn't clear.

But it should be possible to write custom hooks so that they can receive `useState`, `useEffect`, etc. as parameters so that React or Haunted or any other React-compatible hooks library could be used:

**create-my-hook.js:**
```js
export const createMyHook = (useState, useEffect) => {
  // ...
};
```

**my-hook-react.js:**
```js
import { createMyHook } from './create-my-hook.js';
import { useState, useEffect } from 'react';

export const myHook = createMyHook(useState, useEffect);
```

**my-hook-haunted.js:**
```js
import { createMyHook } from './create-my-hook.js';
import { useState, useEffect } from 'haunted';

export const myHook = createMyHook(useState, useEffect);
```

This might be obvious to some, but documenting it up front would help a lot of people, I think. Also, the key to this pattern would be ensuring compatibility for the base hooks.

wdyt?

**Comment by matthewp** [comment]
Yeah, that pattern works. Also anything done to get React libraries to work with Preact should also work, so bundler aliases primarily.

Happy to document this.

**Comment by askbeka** [comment]
Maybe obvious, but there are two approaches at least developers themselves can manage;
replace imports of `react` with `haunted`
Buildtime using bundler to replace  and runtime using "importmap" 

example of later is [here](https://next.plnkr.co/edit/8lukaUKTSzv6HOn5) 

Would be nice to have build time approach and/or runtime approach so that all of [awesome-react-hooks](https://github.com/rehooks/awesome-react-hooks) can be automatically used.

**Comment by askbeka** [comment]
plunkr looks broken, at least for me.

here is the page itself https://run.plnkr.co/plunks/8lukaUKTSzv6HOn5/, but you will have to check for sources in inspector

**Comment by justinfagnani** [comment]
The problem with bundler aliases is that they require a bundler. A factory approach works without tools.

**Comment by askbeka** [comment]
@justinfagnani importmap does not require tools as you can see in plunkr example above. 

But I see your point. **Factory** vs **Dependency Injection**.

**Factory**
Pros:
  - no build tools no runtime `importmap` magic
Cons:
  - consumers should provide all the hooks to the hook library, or authors have to provide separate packages per framework
  - not possible to use available hook libraries like in https://github.com/rehooks/awesome-react-hooks#packages

**DI**
Pros:
 - all hook libraries can be used out of the box;
 - consumers do not need to configure a factory
 - authors do not have to do anything as long as haunted is 1 to 1 compatible with React;

Cons:
- bundler or importmap is required

**Comment by loganvolkers** [comment]
There is an option that provides the best of both **Factory** and **Dependency Injection** would be using a default parameter value.

- [x] Can be used without a bundler
- [x] Can be used with a bundler

**create-my-hook.js:**
```js
import * as Hooks from 'haunted';

export const createMyHook = ({useState, useEffect} = Hooks) => {
  // ...
};
```

**my-hook-react.js:**
```js
import { createMyHook } from './create-my-hook.js';
import * as Hooks from 'react';

export const myHook = createMyHook(Hooks);
```

**my-hook-haunted.js:**
```js
import { createMyHook } from './create-my-hook.js';
import * as Hooks from 'haunted';

export const myHook = createMyHook(Hooks);
```

**Comment by loganvolkers** [comment]
@justinfagnani and @matthewp we have released a library based on my proposal above called `universal-hooks` with the goal of providing a central API across hook implementations. Implementations can be swapped at runtime or build time.

 1) `@saasquatch/universal-hooks` provides an interface which haunted and React can plug into. Write your hooks for the interface and the implementation can be swapped out at runtime or build time.
 2) `@saasquatch/universal-hooks-testing-library`  provides an interface for testing, based off of React Hooks Testing Library. The implementation can be swapped out at runtime or build time.
 3) `@saasquatch/haunted-hooks-testing-library` provides a Universal Hooks compatible testing API for haunted.

This all comes together to make hooks shareable and testable across different hooks implementations.

This dream comes alive in the test suite for `universal-hooks-testing-library`: https://github.com/saasquatch/universal-hooks/blob/main/packages/universal-hooks-testing-library/test/main.ts

Next we'll be focusing on unifying other libraries, such as Preact and [uhooks](https://github.com/WebReflection/uhooks) to this common interface.
